src/blob.jl

```diff
@@ -59,32 +59,30 @@ Defaults to `sizeof(T)`.
     if isempty(fieldnames(T))
         quote
             $(Expr(:meta, :inline))
-            sizeof(T)
+            $(sizeof(T))
         end
     else
         quote
             $(Expr(:meta, :inline))
-            +(0, $(@splice i in 1:length(fieldnames(T)) quote
-                self_size(fieldtype(T, $i))
+            $(+(0, @splice i in 1:length(fieldnames(T)) begin
```

Comment:
Ditto

Reply:
Ditto. Though, it's even worse than the above call, as it results in multiple function calls.

Reply:
Again, I'm not sure I buy this.  With Blobs#master I see this:

`julia> @code_native Blobs.self_size(Int)
	.section	__TEXT,__text,regular,pure_instructions
; Function self_size {
; Location: blob.jl:58
	movl	$8, %eax
	retl
	nopw	%cs:(%eax,%eax)
;}`

Reply:
First, there's an empirical cause and we've seen huge runtime improvement in a user project only because of the change in this PR.
Second, similar to my comment above, there's no guarantee for these kind of optimizations to happen and your example is very simple (just an `Int`).

Here is a little bit more complex example, which also appears in the test cases:

```julia
julia> struct PackedMemoryArray{K,V}
            keys::BlobVector{K}
            values::BlobVector{V}
            mask::BlobBitVector
            count::Int
            #...other stuff
       end

julia> @code_warntype Blobs.self_size(PackedMemoryArray{Int,Int})
Body::Int64
58 1 ─ %1 = (Base.add_int)(32, 16)::Int64               │╻╷╷╷ macro expansion
   │   %2 = (Base.add_int)(%1, 8)::Int64                ││╻    +
   └──      return %2                                   ││   
```

Reply:
I'm just trying to get to the bottom of why this change should help; I feel like there's too much guesswork in the current performance tuning workflow.  The PackedMemoryArray example still compiles away, once llvm does its thing:
```julia-repl
julia> @code_native Blobs.self_size(PackedMemoryArray{Int,Int})
	.section	__TEXT,__text,regular,pure_instructions
; Function self_size {
; Location: blob.jl:58
; Function macro expansion; {
; Location: blob.jl:58
	movl	$56, %eax
	retl
	nopw	%cs:(%eax,%eax)
;}}
```

Reply:
@tjgreen I understand your concern and it's interesting that it's not easy to create a simple example that shows the difference. That's a good sign that Julia compiler works well, at least for the simple examples that I could come up with in a limited amount of time.

However, there's a fundamental difference between the two approaches: 1) generating a constant 2) generating a piece of code that _hopefully_ gets partially evaluated by the compiler. In this case, as we are in a `@generated` function and it's the programmer's responsibility to execute all the code that's possible at code generation time (via `$(...)` or outside `quote ... end`) and only generate code related to the  arguments known at runtime. This is a fact and there is no guesswork that goes with it and was the basis for this change. Though, this fact only tells us this change improves the code quality and removes some burden from the compiler (i.e., less compilation time, both on Julia and LLVM). Your main point was that you wanted a proof for a higher runtime performance, which as I told you earlier, it's based on empirical results and we can definitely find concrete examples if we invest more time on this.